python/lsst/obs/decam/rawFormatter.py

```diff
@@ -22,18 +22,113 @@
+
+        Parameters
+        ----------
+        filename : `str`
+            The file to search through.
+        detectorId : `int`
+            The detector id to search for.
+
+        Returns
+        -------
+        fitsData: `lsst.afw.fits.Fits`
+            An read-only struct for accessing the required HDU.
+        metadata: `lsst.daf.base.PropertyList`
+            The metadata read from the header for that detector id.
+
+        Raises
+        ------
+        ValueError
+            Raised if detectorId is not found in any of the file HDUs
+        """
+        log = lsst.log.Log.getLogger("DarkEnergyCameraRawFormatter")
+        log.debug("Did not find detector=%s at expected HDU=%s in %s: scanning through all HDUs.",
+                  detectorId, detector_to_hdu[detectorId], filename)
+
+        fitsData = lsst.afw.fits.Fits(filename, 'r')
+        # NOTE: The primary header (HDU=0) does not contain detector data.
+        for i in range(1, fitsData.countHdus()):
+            fitsData.setHdu(i)
+            metadata = fitsData.readMetadata()
+            if metadata['CCDNUM'] == detectorId:
+                return fitsData, metadata
+        else:
+            raise ValueError(f"Did not find detectorId={detectorId} as CCDNUM in any HDU of {filename}.")
+
+    def _determineHDU(self, detectorId):
+        """Determine the correct HDU number for a given detector id.
+
+        Parameters
+        ----------
+        detectorId : `int`
+            The detector id to search for.
+
+        Returns
+        -------
+        fitsData : `lsst.afw.fits.Fits`
+            An read-only struct for accessing the required HDU.
+        metadata : `lsst.daf.base.PropertyList`
+            The metadata read from the header for that detector id.
+
+        Raises
+        ------
+        ValueError
+            Raised if detectorId is not found in any of the file HDUs
+        """
+        filename = self.fileDescriptor.location.path
+        try:
+            metadata = lsst.afw.image.readMetadata(filename, detector_to_hdu[detectorId])
+            if metadata['CCDNUM'] != detectorId:
+                # the detector->HDU mapping is different in this file: try scaning
+                return self._scanHdus(filename, detectorId)
+            else:
+                fitsData = lsst.afw.fits.Fits(filename, 'r')
+                fitsData.setHdu(detector_to_hdu[detectorId])
+                return fitsData, metadata
+        except lsst.afw.fits.FitsError:
+            # if the file doesn't contain all the HDUs of "normal" files, try scanning
+            return self._scanHdus(filename, detectorId)
+
+    def readMetadata(self):
+        fitsData, metadata = self._determineHDU(self.dataId['detector'])
+        astro_metadata_translator.fix_header(metadata)
+        return metadata
+
+    def readImage(self):
+        fitsData, metadata = self._determineHDU(self.dataId['detector'])
+        return lsst.afw.image.ImageI(fitsData.readImageI())
```

Comment:
Huh, it looks like those two take exactly the same amount of time. That's a bit surprising to me, because reusing the `afw.fits.Fits` object made the metadata scanning ~3x faster than doing `for i in range(hdus): afw.image.readMetadata(filename, i)`.

In that case, should I not return the `fitsData object from `_determineHDU` and instead just return the hdu index?

Reply:
Yes, it might be that doing the loop you gain because you aren't having to continually reopen the file, whereas reading a single image from a file is dominated by reading the pixels in. Anyhow, yes, return an HDU instead.